feat(import): add `max_candidates` config option

ali-ramadhan opened the pull request:
- [x] I have read the contributing guide in the documentation.

### Description

Introduced a new config option `max_candidates` to control the number of candidate albums displayed during the import process. Updated the import prompt to respect this setting and added validation for the configuration. Included docs and tests.

This option will be useful when using multiple multiple metadata plugins as currently `max_candidates` is hard coded to be 5. The docs mention a `search_limit` config option for the MusicBrainz plugin which is introduced by https://github.com/MoeMusic/moe_musicbrainz/pull/12.

<!-- readthedocs-preview mrmoe start -->
----
📚 Documentation preview 📚: https://mrmoe--302.org.readthedocs.build/en/302/

<!-- readthedocs-preview mrmoe end -->

ali-ramadhan commented:
Hmmm, I didn't see the pyright errors from https://github.com/MoeMusic/Moe/actions/runs/15551356169/job/43782444107?pr=302 locally but will revisit tomorrow to see if I can reproduce locally and fix them.

jtpavlock commented:
It's probably due to the ci pulling in a newer version of pyright. See if you can update your local version with `poetry update` if that's the case. Just make sure to request a review also whenever you're ready so I get the notification.

jtpavlock reviewed (changes requested):
Nice work here yet again, only a couple minor things as this is fairly similar to the search_limit config change. 

Could you also do me a favor and try something new for your commit changes? With the force pushing flow we're doing now, it ends up being difficult to compare changes between reviews. Instead, I'd like to try to use [fixup commits](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt-code--fixupamendrewordltcommitgtcode) for any commit edits after the initial review. Then, I'll go in and squash them before merging into main.

ali-ramadhan commented on moe/moe_import/__init__.py:
```diff
@@ -14,6 +17,14 @@
 
 
+@moe.hookimpl
+def add_config_validator(settings: dynaconf.base.LazySettings):
```
Done! I moved it close to `process_candidates` where the config is used, but let me know if it would make more sense to move the function elsewhere in the file.

ali-ramadhan commented on moe/moe_import/import_cli.py:
```diff
@@ -113,7 +113,8 @@ def process_candidates(new_album: Album, candidates: list[CandidateAlbum]):
     if candidates:
         try:
-            candidate_prompt(new_album, candidates[:5])
+            max_candidates = config.CONFIG.settings.get("import.max_candidates", 5)
```
Sounds good and done!

ali-ramadhan commented on tests/import/test_import_cli.py:
```diff
@@ -136,6 +136,64 @@ def test_process_no_candidates(self):
 
+    def test_max_candidates_default(self):
+        """By default, only the first 5 candidates should be processed."""
+        mock_new_album = Mock()
```
Thanks for the tip! I did not know about `ANY`. I've used it and updated the tests.

ali-ramadhan commented:
Thanks again for the thorough review @jtpavlock! I believe I've addressed your review comments so just requesting another review.

> With the force pushing flow we're doing now, it ends up being difficult to compare changes between reviews. Instead, I'd like to try to use [fixup commits](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt-code--fixupamendrewordltcommitgtcode) for any commit edits after the initial review. Then, I'll go in and squash them before merging into main.

100% agree with it being difficult to review changes. Just made a `fixup!` commit! I am a big fan of squash and merge although I'm usually squashing a pretty dirty history with merge commits and "fix tests" commits haha. Thanks for pointing out the `fixup` commit feature. Will use them in future PRs and will try using them in other projects.

jtpavlock commented on tests/import/test_import_cli.py:
```diff
@@ -164,35 +161,44 @@ def test_max_candidates_custom(self, tmp_config):
+                [import]
+                max_candidates = 0
+                """,
+                tmp_db=True,
```
We don't need to create a database here (same for test_max_candidates_option) so this can be removed and it should speed up the test a bit.

jtpavlock commented on tests/import/test_import_cli.py:
```diff
@@ -164,35 +161,44 @@ def test_max_candidates_custom(self, tmp_config):
         ) as mock_candidate_prompt:
             config.CONFIG.pm.hook.process_candidates(
-                new_album=mock_new_album,
+                new_album=ANY,
```
Sorry for leading you astray here, but `ANY` should only be used for assertions, I don't think it's right to use it for passing values. `new_album=Mock()` would be more appropriate, but leave `ANY` in the assertion check. Same goes for the other tests.

ali-ramadhan replied:
Ah thanks for the explanation! This is also pointed out in the docs: https://docs.python.org/3/library/unittest.mock.html#any

Fixed!

jtpavlock reviewed (changes requested):
Very minor changes then I'll merge it in. The `fixup!` flow also was much easier to see the changes, thanks! This is my first time trying it out, but I like it already as it removes the need to force push and you can make as many `fixup!` commits as you want. 

For now I'll have to manually squash them, but in the future, hopefully [github can do the squashing too](https://github.com/orgs/community/discussions/40804).

jtpavlock commented on moe/moe_import/import_cli.py:
```diff
@@ -108,12 +110,21 @@ def add_import_prompt_choice(prompt_choices: list[PromptChoice]):
 def process_candidates(new_album: Album, candidates: list[CandidateAlbum]):
     """Use the import prompt to select and process the imported candidate albums."""
     if candidates:
+        max_candidates = config.CONFIG.settings.get("import.max_candidates")
```
.get() isn't needed just `settings.import.max_candidates`

ali-ramadhan replied:
Fixed!

ali-ramadhan replied:
Ah so black, flake8, and pyright did not like this, probably because `import` is a reserved keyword. So I'll keep this one as-is.

```
❯ poetry run pre-commit run -a
black....................................................................Failed
- hook id: black
- exit code: 123

All done! ✨ 🍰 ✨4 files left unchanged.
error: cannot format moe/moe_import/import_cli.py: Cannot parse: 125:48:         max_candidates = config.CONFIG.settings.import.max_candidates

Oh no! 💥 💔 💥3 files left unchanged, 1 file failed to reformat.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨4 files left unchanged.
All done! ✨ 🍰 ✨3 files left unchanged.

flake8...................................................................Failed
- hook id: flake8
- exit code: 1

moe/moe_import/import_cli.py:125:50: E999 SyntaxError: invalid syntax

isort....................................................................Passed
pyright..................................................................Failed
- hook id: pyright
- exit code: 1

0 errors, 0 warnings, 0 informations 
/home/alir/code/Moe/moe/moe_import/import_cli.py
  /home/alir/code/Moe/moe/moe_import/import_cli.py:125:48 - error: Expected attribute name after "."
  /home/alir/code/Moe/moe/moe_import/import_cli.py:125:49 - error: Statements must be separated by newlines or semicolons
  /home/alir/code/Moe/moe/moe_import/import_cli.py:125:55 - error: Relative imports cannot be used with "import .a" form; use "from . import a" instead
  /home/alir/code/Moe/moe/moe_import/import_cli.py:125:55 - error: Import ".max_candidates" could not be resolved (reportMissingImports)
4 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations 
0 errors, 0 warnings, 0 informations
```

ali-ramadhan commented:
> For now I'll have to manually squash them, but in the future, hopefully [github can do the squashing too](https://github.com/orgs/community/discussions/40804).

Ah I was wondering what the difference is between regular "squash and merge" or "rebase and merge" and this but it sounds like with `fixup!` commits it's guaranteed that you can "squash, rebase, and then merge"?

jtpavlock commented:
So I guess the main difference is squashing creates a new commit whereas rebasing maintains the original commit. I also like the flexibility with rebasing to have more than one commit in a pull request, such as a feat and a refactor or several feats, etc.. If the PR only has one commit I think squash merging and rebasing have the same practical end result of moving a single commit onto main though.

Also, you shouldn't have to be force pushing your branch with these fixup commits, is it giving you issues with a normal push?

jtpavlock reviewed (approved):
thanks for the changes, this looks good

jtpavlock commented:
Oh okay I see what you're saying. I've never actually run into that nor did I consider it a potential issue haha. It definitely seems like good practice though, you're right though that I'm not a fan of merge commits and do prefer rebasing so I think it makes sense to keep doing what you're doing. Thanks for the explanation.
